mypy/checker.py

```diff
@@ -2756,3 +2749,51 @@ def is_valid_inferred_type_component(typ: Type) -> bool:
             if not is_valid_inferred_type_component(item):
                 return False
     return True
+
+
+class Scope:
+    # We keep two stacks combined, to maintain the relative order
+    stack = None  # type: List[Union[Type, FuncItem, MypyFile]]
+
+    def __init__(self, module: MypyFile) -> None:
+        self.stack = [module]
+
+    def push_function(self, item: FuncItem) -> None:
+        self.stack.append(item)
+
+    def pop_function(self) -> None:
+        assert isinstance(self.stack[-1], Type)
+        self.stack.pop()
+
+    def top_function(self) -> Optional[FuncItem]:
+        for e in reversed(self.stack):
+            if isinstance(e, FuncItem):
+                return e
+        return None
+
+    def push_class(self, t: Type) -> None:
+        self.stack.append(t)
+
+    def pop_class(self) -> None:
+        assert isinstance(self.stack[-1], Type)
+        self.stack.pop()
+
+    def active_class(self) -> Optional[Type]:
+        if isinstance(self.stack[-1], Type):
+            return self.stack[-1]
+        return None
+
+    @contextmanager
+    def enter_function(self, item: FuncItem) -> Iterator[None]:
```

Comment:
Shouldn't this use push/pop_function() to make it clear it's doing the right thing? Ditto enter_class().